Memcached unix socket panic

I have configured memcached as

```
# Read metrics from one or many memcached servers
[[inputs.memcached]]
  ## An array of address to gather stats about. Specify an ip on hostname
  ## with optional port. ie localhost, 10.0.0.1:11211, etc.
  # servers = ["localhost:11211"]
  unix_sockets = ["/var/run/memcached/memcached.sock"]
```

But this throws the following panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x40 pc=0x56e61b]

goroutine 1 [running]:
github.com/influxdata/telegraf/plugins/inputs/memcached.(*Memcached).gatherServer(0xc8200c3950, 0xc8200c38c1, 0x21, 0xc8200c8101, 0x7f4120b25340, 0xc82004af00, 0x0, 0x0)
        /home/ubuntu/telegraf-build/src/github.com/influxdata/telegraf/plugins/inputs/memcached/memcached.go:117 +0x1fb
github.com/influxdata/telegraf/plugins/inputs/memcached.(*Memcached).Gather(0xc8200c3950, 0x7f4120b25340, 0xc82004af00, 0x0, 0x0)
        /home/ubuntu/telegraf-build/src/github.com/influxdata/telegraf/plugins/inputs/memcached/memcached.go:83 +0x261
github.com/influxdata/telegraf/agent.(*Agent).Test(0xc820028000, 0x0, 0x0)
        /home/ubuntu/telegraf-build/src/github.com/influxdata/telegraf/agent/agent.go:228 +0x434
main.main()
        /home/ubuntu/telegraf-build/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:190 +0x120e
```

The file `/var/run/memcached/memcached.sock` exists and has 766 memcached:memcached permissions.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,7 @@
 - [#898](https://github.com/influxdata/telegraf/issues/898): Put database name in quotes, fixes special characters in the database name.
 - [#656](https://github.com/influxdata/telegraf/issues/656): No longer run `lsof` on linux to get netstat data, fixes permissions issue.
 - [#907](https://github.com/influxdata/telegraf/issues/907): Fix prometheus invalid label/measurement name key.
+- [#841](https://github.com/influxdata/telegraf/issues/841): Fix memcached unix socket panic.
 
 ## v0.11.1 [2016-03-17]
 

--- a/plugins/inputs/memcached/memcached.go
+++ b/plugins/inputs/memcached/memcached.go
@@ -94,14 +94,15 @@ func (m *Memcached) gatherServer(
 	acc telegraf.Accumulator,
 ) error {
 	var conn net.Conn
+	var err error
 	if unix {
-		conn, err := net.DialTimeout("unix", address, defaultTimeout)
+		conn, err = net.DialTimeout("unix", address, defaultTimeout)
 		if err != nil {
 			return err
 		}
 		defer conn.Close()
 	} else {
-		_, _, err := net.SplitHostPort(address)
+		_, _, err = net.SplitHostPort(address)
 		if err != nil {
 			address = address + ":11211"
 		}
@@ -111,6 +112,10 @@ func (m *Memcached) gatherServer(
 			return err
 		}
 		defer conn.Close()
+	}
+
+	if conn == nil {
+		return fmt.Errorf("Failed to create net connection")
 	}
 
 	// Extend connection

